File: installer/payload.py
"""Locate the installable artifacts — the light OBJs, the textures, the plugin,
and the precomputed RealWings patch data.

**ONE MODE: bundled.** Read the pre-built `payload/` directory that sits next to
`install.py`, resolved relative to this file — never CWD, so the installer runs
from any directory (Downloads, a USB stick, or inside
X-Plane/Resources/plugins/). A frozen **loose** snapshot instead reads a `data/`
folder next to the executable (same layout) — see `_default_payload_dir`.

⚠ THE DEV MODE IS GONE, DELIBERATELY (docs/installer_cpp_plan.md §2b). Running
`install.py` from a repo checkout with no `payload/` used to fall back to the
repo's own build system and generate each OBJ on the fly via `build_objs`. That
made the DSL toolchain an INSTALL-TIME dependency, which a compiled installer
cannot honor, and it meant the two paths could drift. The dev path is now an
explicit staging step instead:

    python build/make_release.py --payload-only     # stage payload/ beside install.py
    python install.py

Bundled `payload/` layout (built by build/make_release.py):
    payload/objs/<stock|durantula|realwings>/<obj filename>
    payload/objs/interior/lights_inn.obj
    payload/objs/screens/lights_screens.obj
    payload/textures/interior/*.png                       (Gus's set, ~57 MiB)
    payload/plugin/ToLissPhoton/<arch>/ToLissPhoton.xpl   (arch = win_x64|mac_x64|lin_x64)
    payload/plugindata/{panelfx.txt,overlays/*.png}       (into the plugin folder)
    payload/realwings_patch.json                          (resolved from the DSL)

⚠ There is NO `payload/dsl/` any more. The RealWings live patch was the last
install-time DSL consumer; its numbers are now resolved into
`realwings_patch.json` at bundle time and applied by `installer/realwings.py`.

The plugin is the compiled native `.xpl` (no XPPython3/Python at runtime). It is a
fat plugin: the folder can carry one `.xpl` per platform side by side, and X-Plane
loads the subfolder matching the host OS.
"""
from __future__ import annotations

import sys
from pathlib import Path

from installer import realwings
from installer.constants import (
    AIRFRAMES, INTERIOR_OBJ, INTERIOR_TEXTURES, PANELFX_FILE, PLUGIN_DATA_DIRNAME,
    PLUGIN_FOLDER, PLUGIN_USER_DIRS, SCREENS_OBJ, WINGS_FOR, XPL_NAME,
)

# installer/payload.py -> installer/ -> the dir holding install.py. In a release
# that dir holds payload/; in the repo it *is* the repo root (holds build/, src/).
ROOT = Path(__file__).resolve().parent.parent


def _default_payload_dir() -> Path:
    """Where the installable artifacts live by default.

    A frozen **loose** snapshot (`build/make_exe.py --loose`) ships the installer
    and its files as separate items: `<installer>.exe` beside a plainly-named
    `data/` folder. When such a folder sits next to the executable it wins, so the
    exe reads the real files on disk (the tester can see/swap them). Absent that,
    fall back to `ROOT/payload`: for a frozen **self-contained** exe that is the
    copy PyInstaller embedded under its extraction dir; running from source (or
    under tests, which repoint PAYLOAD_DIR at a fixture) it is the repo/release
    root. `getattr(sys, "frozen", ...)` is only ever True in a PyInstaller build,
    so source and test runs always take the `ROOT/payload` branch unchanged."""
    if getattr(sys, "frozen", False):
        loose = Path(sys.executable).resolve().parent / "data"
        if loose.is_dir():
            return loose
    return ROOT / "payload"


PAYLOAD_DIR = _default_payload_dir()


class PayloadError(Exception):
    """The artifacts couldn't be located or built — a broken/incomplete release
    or a repo checkout missing its build inputs, not a user-fixable condition."""


def _read_obj(path: Path) -> str:
    """Read a bundled OBJ WITHOUT newline translation.

    ⚠ `newline=""` IS LOAD-BEARING. `build_objs.Emitter.emit()` ends with
    `.replace("\\n", "\\r\\n")` — X-Plane's OBJs are CRLF and ours match — so a
    plain `read_text()` universal-newlines the whole file to LF on the way in, and
    the installer then writes an LF copy of a CRLF artifact. Nothing breaks (the
    parser takes either), which is exactly why it went unnoticed: the installed
    file simply stopped being byte-identical to the payload, so any byte-level
    comparison against it — a diff, a checksum, the C++/Python parity harness —
    reported a whole-file difference with no cause visible in either tree.

    Found by `tests/test_parity.py` (2026-08-07); same rule the `.acf` patchers
    already follow for the same reason."""
    return path.read_text(encoding="utf-8", errors="replace", newline="")


# ─── availability ──────────────────────────────────────────────────────────────
def _bundled() -> bool:
    return PAYLOAD_DIR.is_dir()


def available() -> bool:
    return _bundled()


def mode() -> str:
    return "bundled" if _bundled() else "unavailable"


def _require_bundle() -> None:
    """The one place that explains a missing bundle, including the dev recipe —
    since the fix for a repo checkout is no longer "it builds itself"."""
    if _bundled():
        return
    raise PayloadError(
        f"no bundled payload at {PAYLOAD_DIR} — this build is incomplete. "
        f"From a repo checkout, stage one first: "
        f"python build/make_release.py --payload-only")


# ─── OBJs ──────────────────────────────────────────────────────────────────────
def obj_path(wing: str, airframe: str) -> Path:
    """Path to a pre-built OBJ in the bundle."""
    _, _, fname, _ = AIRFRAMES[airframe]
    p = PAYLOAD_DIR / "objs" / wing / fname
    if not p.is_file():
        raise PayloadError(f"missing bundled OBJ: {p}")
    return p


def obj_text(wing: str, airframe: str) -> str:
    """The OBJ text to install, read from the bundle.

    Checked against WINGS_FOR first even though a missing file would already fail
    in `obj_path`: the message "a339 has no 'durantula' variant" is the true
    reason, where "missing bundled OBJ" reads as a broken download. `make_release`
    applies the same gate when building the payload (SUPPORTED_WINGS), so the two
    ends agree on which combinations exist at all."""
    if wing not in WINGS_FOR[airframe]:
        raise PayloadError(f"{airframe} has no {wing!r} wing-mod variant "
                           f"(supported: {', '.join(WINGS_FOR[airframe])})")
    _require_bundle()
    return _read_obj(obj_path(wing, airframe))


# ─── interior OBJ + textures ("Gus Mod") ───────────────────────────────────────
# The interior artifact is ONE file for everything: no wing variants (wing mods
# don't touch the cockpit) and no per-airframe variants (stock lights_inn.obj is
# byte-identical across A319/A320/A321). Hence no arguments — a per-(wing,
# airframe) signature like obj_text's would imply a choice that doesn't exist,
# and in the bundle all three airframes would resolve to the same path anyway.
def interior_obj_path() -> Path:
    p = PAYLOAD_DIR / "objs" / "interior" / INTERIOR_OBJ
    if not p.is_file():
        raise PayloadError(f"missing bundled interior OBJ: {p}")
    return p


def interior_obj_text() -> str:
    """The interior OBJ text to install. Built from the a320 fixtures, which
    a319/a321 inherit verbatim (a339 has no interior target at all)."""
    _require_bundle()
    return _read_obj(interior_obj_path())


# ─── screen-glow OBJ ───────────────────────────────────────────────────────────
# Same "one file for everything" shape as the interior OBJ, and for the same
# reasons: no wing variants (wing mods don't touch the cockpit) and no per-airframe
# variants (the A3xx flight decks are the same model, so the six display positions
# are identical). Hence no arguments. a339 has no `screens` target at all.
def screens_obj_path() -> Path:
    p = PAYLOAD_DIR / "objs" / "screens" / SCREENS_OBJ
    if not p.is_file():
        raise PayloadError(f"missing bundled screen-glow OBJ: {p}")
    return p


def screens_obj_text() -> str:
    """The screen-glow OBJ text to install.

    ⚠ It must never be a `--debug` build. That variant is a tuning artifact bound
    to `ToLissPhoton/debug/light/*`, is NOT installable, and would leave six lights
    ignoring the DU knobs. `make_release.build_screens_payload` emits it straight
    from the Emitter for exactly this reason, rather than copying `dist/` — which
    on a dev machine may well hold a debug build."""
    _require_bundle()
    return _read_obj(screens_obj_path())


def screens_available() -> bool:
    """Whether this build can install the screen glow at all. Unlike the interior
    this is not a user opt-in, so a False here just drops the step rather than
    hiding a menu — an older bundle with no `objs/screens/` installs everything
    else and says so."""
    try:
        screens_obj_text()
        return True
    except (PayloadError, OSError):
        return False


def interior_texture_dir() -> Path:
    """Where Gus's canonical texture set lives in the bundle. `make_release`
    stages it from the vendored `reference/gus/textures/` (the build must NOT
    depend on `.scratch/`, which is gitignored and absent from a fresh clone)."""
    return PAYLOAD_DIR / "textures" / "interior"


def interior_textures() -> list[tuple[str, Path]]:
    """The texture set as (filename, absolute source). Ordered by name so the
    install progress display is stable run to run.

    Only files actually present are returned, but a MISSING one is an error, not
    a silent skip: shipping 9 of 11 would leave the cockpit half-retextured."""
    src = interior_texture_dir()
    if not src.is_dir():
        raise PayloadError(
            f"interior texture set not found: {src} — this build is incomplete "
            f"(see reference/gus/README.md)")
    out, missing = [], []
    for name in INTERIOR_TEXTURES:
        p = src / name
        (out if p.is_file() else missing).append((name, p) if p.is_file() else name)
    if missing:
        raise PayloadError(
            f"interior texture set is incomplete in {src} — missing: "
            + ", ".join(missing))
    return out


def interior_available() -> bool:
    """Whether this build can install the interior mod at all, so the TUI can
    hide the opt-in rather than offer a step that would fail."""
    try:
        interior_obj_text()
        interior_textures()
        return True
    except (PayloadError, OSError):
        return False


# ─── the RealWings patch data ──────────────────────────────────────────────────
# ⚠ Resolved from the DSL at BUNDLE time (build/patch_realwings.py) and applied at
# install time (installer/realwings.py). This is the file that replaced
# `payload/dsl/` — see that module's header for why the split exists.
def realwings_patch_path() -> Path:
    return PAYLOAD_DIR / realwings.PATCH_JSON


def realwings_patch_data() -> dict:
    """The resolved RealWings replacement parameters.

    Loaded per call rather than cached: PAYLOAD_DIR is repointed by tests, and the
    file is read at most once per install."""
    return realwings.load(realwings_patch_path())


def realwings_available() -> bool:
    """Whether this bundle can patch the RealWings mod at all. A bundle built
    before the split has no such file; the install then continues without the
    wingtip patch rather than refusing to run."""
    try:
        realwings_patch_data()
        return True
    except realwings.RealWingsError:
        return False


# ─── plugin (native fat-plugin folder) ──────────────────────────────────────────
def plugin_src_root() -> Path:
    """The `ToLissPhoton/` fat-plugin folder to install, from the bundle."""
    p = PAYLOAD_DIR / "plugin" / PLUGIN_FOLDER
    if not p.is_dir():
        raise PayloadError(f"native plugin folder not found: {p} — "
                           f"this release bundle is incomplete")
    return p


def plugin_files() -> list[tuple[str, Path]]:
    """Every file in the fat-plugin folder as (folder-relative POSIX path, absolute
    source) — normally `<arch>/ToLissPhoton.xpl` for each built platform. Copied
    verbatim into <X-Plane>/Resources/plugins/ToLissPhoton/ at install time.

    Deliberately a whole-tree walk rather than a hard-coded `<arch>/<XPL_NAME>`
    list, so a future runtime data file ships without touching this — except for
    PLUGIN_USER_DIRS, which is user territory (`overlays/`). A source folder that
    happens to be a *deployed* plugin folder would otherwise carry a dev machine's
    overlay images into every install."""
    root = plugin_src_root()
    files = [(rel, p) for rel, p in
             ((p.relative_to(root).as_posix(), p) for p in sorted(root.rglob("*")))
             if p.is_file() and not _is_user_owned(rel)]
    if not files:
        raise PayloadError(f"native plugin folder is empty (no .xpl built?): {root}")
    return files


def _is_user_owned(rel: str) -> bool:
    """Whether a plugin-folder-relative POSIX path lies under a PLUGIN_USER_DIRS
    folder — matched on the first path SEGMENT, so a `.xpl` named `overlays.xpl`
    is not caught by a prefix test."""
    head = rel.split("/", 1)[0]
    return head in PLUGIN_USER_DIRS


def plugin_data_root() -> Path | None:
    """The tree of runtime DATA files that go into the plugin folder alongside the
    `.xpl` — `panelfx.txt` and `overlays/` — staged at `payload/plugindata/`.

    None, not an error, when it does not exist: a bundle built before the
    compositor shipped installs the plugin and nothing else, rather than refusing
    to run."""
    p = PAYLOAD_DIR / PLUGIN_DATA_DIRNAME
    return p if p.is_dir() else None


def plugin_data_files() -> list[tuple[str, Path]]:
    """The shipped plugin-folder data files as (folder-relative POSIX path,
    absolute source): `panelfx.txt` plus every image under `overlays/`.

    ⚠ Enumerated from a fixed set of ROOTS, not from a whole-tree walk — the roots
    are what `make_release.stage_plugin_data` writes, and keeping the two on the
    same shape is what stops a stray file in `plugindata/` becoming an installed
    artifact. `plugin_files()` can walk its root because that root is only ever a
    built plugin folder; this one is a staging directory."""
    root = plugin_data_root()
    if root is None:
        return []
    out: list[tuple[str, Path]] = []
    fx = root / PANELFX_FILE
    if fx.is_file():
        out.append((PANELFX_FILE, fx))
    for d in PLUGIN_USER_DIRS:
        src = root / d
        if not src.is_dir():
            continue
        for f in sorted(src.rglob("*")):
            if f.is_file():
                out.append((f"{d}/{f.relative_to(src).as_posix()}", f))
    return out


def plugin_arches() -> list[str]:
    """Platform arch subfolders that actually contain a `.xpl` (win_x64/mac_x64/
    lin_x64), for reporting which platforms this bundle ships."""
    root = plugin_src_root()
    return sorted(d.name for d in root.iterdir()
                  if d.is_dir() and (d / XPL_NAME).is_file())

File: src/native/src/core/payload.cpp
#include "core/payload.h"

#include <algorithm>
#include <sstream>

#include "core/constants.h"
#include "core/fsutil.h"

namespace photon {
namespace payload {
namespace {

namespace fs = std::filesystem;

std::string gPayloadDir;

std::string Join(const std::vector<std::string>& v, const char* sep) {
    std::ostringstream os;
    for (std::size_t i = 0; i < v.size(); ++i) {
        if (i) os << sep;
        os << v[i];
    }
    return os.str();
}

fs::path Dir() { return fsutil::PathFromUtf8(gPayloadDir); }

// The one place that explains a missing bundle, INCLUDING the dev recipe — since
// the fix for a repo checkout is no longer "it builds itself".
void RequireBundle() {
    if (Available()) return;
    throw PayloadError(
        "no bundled payload at " + gPayloadDir +
        " — this build is incomplete. From a repo checkout, stage one first: "
        "python build/make_release.py --payload-only");
}

std::string ReadTextOrThrow(const fs::path& p, const char* what) {
    std::string text;
    if (!fsutil::ReadFileBytes(p, text)) {
        throw PayloadError(std::string("missing bundled ") + what + ": " +
                           fsutil::PathToUtf8(p));
    }
    return text;
}

}  // namespace

void InitFromExecutable(const std::string& exePathUtf8) {
    std::error_code ec;
    fs::path exe = fs::weakly_canonical(fsutil::PathFromUtf8(exePathUtf8), ec);
    if (ec) exe = fsutil::PathFromUtf8(exePathUtf8);
    const fs::path dir = exe.parent_path();

    // ⚠ TWO NAMES, ONE BINARY. `payload/` is what `make_release.py` stages in a
    // checkout and the canonical name. `data/` is what the shipped per-platform
    // bundle calls it — the plainer end-user name, documented in its README — and
    // this same executable has to run out of both: a dev restages `payload/` and
    // reruns, a user downloads `data/`. Renaming either would break the other for
    // no gain; a second accepted name is one line.
    //
    // Order matters: `payload/` wins, so a repo checkout that has staged one is
    // never shadowed by a stale `data/` left over from a bundle build.
    gPayloadDir = fsutil::PathToUtf8(dir / "payload");
    if (!Available()) {
        const std::string alt = fsutil::PathToUtf8(dir / "data");
        const std::string canonical = gPayloadDir;
        gPayloadDir = alt;
        if (!Available()) gPayloadDir = canonical;   // report the canonical name
    }
}

void SetPayloadDir(const std::string& dirUtf8) { gPayloadDir = dirUtf8; }

std::string PayloadDir() { return gPayloadDir; }

bool Available() {
    std::error_code ec;
    return !gPayloadDir.empty() && fs::is_directory(Dir(), ec) && !ec;
}

std::string ObjPath(const std::string& wing, const std::string& airframeKey) {
    const Airframe* af = AirframeByKey(airframeKey);
    if (af == nullptr) throw PayloadError("unknown airframe: " + airframeKey);
    const fs::path p = Dir() / "objs" / fsutil::PathFromUtf8(wing) / af->objName;
    std::error_code ec;
    if (!fs::is_regular_file(p, ec) || ec) {
        throw PayloadError("missing bundled OBJ: " + fsutil::PathToUtf8(p));
    }
    return fsutil::PathToUtf8(p);
}

std::string ObjText(const std::string& wing, const std::string& airframeKey) {
    // ⚠ BEFORE the bundle check — see the header. "a339 has no 'durantula'
    // variant" is the true reason; "this build is incomplete" is a wrong one that
    // sends the user to re-download.
    if (!WingIsOfferedFor(airframeKey, wing)) {
        throw PayloadError(airframeKey + " has no '" + wing +
                           "' wing-mod variant (supported: " +
                           Join(WingsFor(airframeKey), ", ") + ")");
    }
    RequireBundle();
    return ReadTextOrThrow(fsutil::PathFromUtf8(ObjPath(wing, airframeKey)), "OBJ");
}

std::string InteriorObjPath() {
    const fs::path p = Dir() / "objs" / "interior" / kInteriorObj;
    std::error_code ec;
    if (!fs::is_regular_file(p, ec) || ec) {
        throw PayloadError("missing bundled interior OBJ: " + fsutil::PathToUtf8(p));
    }
    return fsutil::PathToUtf8(p);
}

std::string InteriorObjText() {
    RequireBundle();
    return ReadTextOrThrow(fsutil::PathFromUtf8(InteriorObjPath()), "interior OBJ");
}

std::string ScreensObjPath(const std::string& airframeKey) {
    // ⚠ The airframe gate fires BEFORE the file check, the same way ObjPath's wing
    // gate does and for the same reason: an airframe this feature does not cover
    // must report "not supported here", never "your download is incomplete".
    if (!AirframeHasScreens(airframeKey)) {
        throw PayloadError("the display-glow OBJ is not supported on " +
                           airframeKey);
    }
    const fs::path p = Dir() / "objs" / "screens" / airframeKey / kScreensObj;
    std::error_code ec;
    if (!fs::is_regular_file(p, ec) || ec) {
        throw PayloadError("missing bundled screen-glow OBJ: " +
                           fsutil::PathToUtf8(p));
    }
    return fsutil::PathToUtf8(p);
}

std::string ScreensObjText(const std::string& airframeKey) {
    RequireBundle();
    return ReadTextOrThrow(fsutil::PathFromUtf8(ScreensObjPath(airframeKey)),
                           "screen-glow OBJ");
}

bool ScreensAvailable(const std::string& airframeKey) {
    try {
        (void)ScreensObjPath(airframeKey);
        return Available();
    } catch (const PayloadError&) {
        return false;
    }
}

std::vector<std::pair<std::string, std::string>> InteriorTextureFiles() {
    const fs::path src = Dir() / "textures" / "interior";
    std::error_code ec;
    if (!fs::is_directory(src, ec) || ec) {
        throw PayloadError("interior texture set not found: " +
                           fsutil::PathToUtf8(src) +
                           " — this build is incomplete");
    }
    std::vector<std::pair<std::string, std::string>> out;
    std::vector<std::string> missing;
    for (const std::string& name : InteriorTextures()) {
        const fs::path p = src / fsutil::PathFromUtf8(name);
        std::error_code fe;
        if (fs::is_regular_file(p, fe) && !fe) {
            out.emplace_back(name, fsutil::PathToUtf8(p));
        } else {
            missing.push_back(name);
        }
    }
    if (!missing.empty()) {
        // ⚠ An error, not a silent skip: shipping 9 of 11 leaves the cockpit
        // half-retextured, which reads in-sim as the mod being broken.
        throw PayloadError("interior texture set is incomplete in " +
                           fsutil::PathToUtf8(src) + " — missing: " +
                           Join(missing, ", "));
    }
    return out;
}

bool InteriorAvailable() {
    try {
        (void)InteriorObjPath();
        (void)InteriorTextureFiles();
        return Available();
    } catch (const PayloadError&) {
        return false;
    }
}

std::string RealWingsPatchPath() {
    return fsutil::PathToUtf8(Dir() / patch_realwings::kPatchJson);
}

patch_realwings::PatchData RealWingsPatchData() {
    return patch_realwings::Load(RealWingsPatchPath());
}

bool RealWingsAvailable() {
    try {
        (void)RealWingsPatchData();
        return true;
    } catch (const patch_realwings::RealWingsError&) {
        return false;
    }
}

std::string PluginSrcRoot() {
    const fs::path p = Dir() / "plugin" / kPluginFolder;
    std::error_code ec;
    if (!fs::is_directory(p, ec) || ec) {
        throw PayloadError("native plugin folder not found: " +
                           fsutil::PathToUtf8(p) +
                           " — this release bundle is incomplete");
    }
    return fsutil::PathToUtf8(p);
}

std::vector<std::pair<std::string, std::string>> PluginFiles() {
    const fs::path root = fsutil::PathFromUtf8(PluginSrcRoot());
    std::vector<std::pair<std::string, std::string>> out;
    for (const fsutil::FileEntry& e : fsutil::ListFilesRecursive(root)) {
        // ⚠ `overlays/` is the user's territory. A source folder that happens to be
        // a DEPLOYED plugin folder holds that dev's own images.
        if (PluginPathIsUserOwned(e.rel)) continue;
        out.emplace_back(e.rel, fsutil::PathToUtf8(e.abs));
    }
    if (out.empty()) {
        throw PayloadError("native plugin folder is empty (no .xpl built?): " +
                           fsutil::PathToUtf8(root));
    }
    return out;
}

std::vector<std::pair<std::string, std::string>> PluginDataFiles() {
    std::vector<std::pair<std::string, std::string>> out;
    const fs::path root = Dir() / kPluginDataDirName;
    std::error_code ec;
    // None, not an error: a bundle built before the compositor shipped installs
    // the plugin and nothing else, rather than refusing to run.
    if (!fs::is_directory(root, ec) || ec) return out;

    const fs::path fx = root / kPanelFxFile;
    std::error_code fe;
    if (fs::is_regular_file(fx, fe) && !fe) {
        out.emplace_back(kPanelFxFile, fsutil::PathToUtf8(fx));
    }
    for (const std::string& d : PluginUserDirs()) {
        const fs::path sub = root / fsutil::PathFromUtf8(d);
        std::error_code de;
        if (!fs::is_directory(sub, de) || de) continue;
        for (const fsutil::FileEntry& e : fsutil::ListFilesRecursive(sub)) {
            out.emplace_back(d + "/" + e.rel, fsutil::PathToUtf8(e.abs));
        }
    }
    return out;
}

std::vector<std::string> PluginArches() {
    const fs::path root = fsutil::PathFromUtf8(PluginSrcRoot());
    std::vector<std::string> out;
    std::error_code ec;
    for (fs::directory_iterator it(root, ec), end; it != end; it.increment(ec)) {
        if (ec) break;
        std::error_code de;
        if (!it->is_directory(de) || de) continue;
        std::error_code fe;
        if (fs::is_regular_file(it->path() / kXplName, fe) && !fe) {
            out.push_back(fsutil::PathToUtf8(it->path().filename()));
        }
    }
    std::sort(out.begin(), out.end());
    return out;
}

}  // namespace payload
}  // namespace photon
